src/main/java/org/folio/rest/impl/TenantPermsAPI.java

```diff
@@ -62,19 +77,396 @@ public void postTenantpermissions(OkapiPermissionSet entity, Map<String, String>
+      Context vertxContext, String tenantId) {
+    Promise<List<Permission>> promise = Promise.promise();
+    
+    try {
+      getPermsByModule(moduleId, vertxContext, tenantId)
+        .onComplete(getResult -> {
+          if (getResult.failed()) {
+            Throwable cause = getResult.cause();
+            logger.error(cause.getMessage(), cause);
+            promise.fail(cause);
+            return;
+          }
+
+          List<Permission> existing = getResult.result();
+          if (existing.isEmpty() && permSet != null) {
+            // this means either:
+            // A) the first time enabling this module, or
+            // B) the permissions exist but don't yet have the definedBy field
+            List<OkapiPermission> perms = permSet.stream()
+                .filter(p -> p.getPermissionName() != null)
+                .collect(Collectors.toList());
+            List<Permission> ret = new ArrayList<>();
+
+            List<Future> futures = new ArrayList<>(perms.size());
+            perms.forEach(perm -> {
+                futures.add(getModulePermByName(perm.getPermissionName(), moduleId.getProduct(),
+                    vertxContext, tenantId).compose(dbPerm -> {
+                  if (dbPerm != null && comparePerms(perm, dbPerm)) {
+                    // (B) we have a match, but lack definedBy
+                    ret.add(dbPerm);
+                  }
+                  return Future.succeededFuture();
+                }));
+            });
+            CompositeFuture.all(futures).onComplete(ar -> {
+              if(ar.failed()) {
+                promise.fail(ar.cause());
+                return;
+              }
+              promise.complete(ret);
+            });
+          } else {
+            promise.complete(existing); // Happy path.
+          }
+        });
+    } catch (Exception e) {
+      logger.error(e.getMessage(), e);
+      promise.fail(e);
+    }
+    
+    return promise.future();
+  }
+
+  private List<OkapiPermission> getNewPerms(Map<String, Permission> dbPerms,
+      List<OkapiPermission> perms) {
+    List<OkapiPermission> newPerms = new ArrayList<>();
+    
+    if(perms != null) {
+      perms.stream()
+        .filter(perm -> perm.getPermissionName() != null)
+        .filter(perm -> !dbPerms.containsKey(perm.getPermissionName()))
+        .filter(perm -> {
+          // filter out renamed perms, we'll deal with them separately
+          for (String oldName : perm.getRenamedFrom()) {
+            if (dbPerms.containsKey(oldName)) {
+              return false;
+            }
+          }
+          return true;
+        })
+        .forEach(newPerms::add);
+    }
+
+    return newPerms;
+  }
+
+  private Map<OkapiPermission, Permission> getRenamedPerms(Map<String, Permission> dbPerms,
+      List<OkapiPermission> perms) {
+    Map<OkapiPermission, Permission> renamedPerms = new HashMap<>();
```

Comment:
Consider to handle the case where more than one old perms are renamed/replaced by a common new perm

Reply:
interesting - I have to think about this one.  It had not occurred to me.

Reply:
addressed this, and updated the tests to exercise it.